src/zkevm_specs/sig_circuit.py

```diff
@@ -0,0 +1,120 @@
+
+
+class Row:
+    """
+    Signature circuit
+    Verify a message hash is signed by an Ethereum Address.
+    """
+
+    msg_hash: Word
+    sig_v: FQ
+    sig_r: Word
+    sig_s: Word
+    recovered_addr: FQ
+    is_valid: FQ
+
+    ecdsa_chip: ECDSAVerifyChip
+    pub_key_hash: bytes
+    pub_key_x_bytes: bytes
+    pub_key_y_bytes: bytes
+    msg_hash_bytes: bytes
+
+    def __init__(
+        self,
+        pub_key_hash: bytes,
+        address: FQ,
+        msg_hash: Word,
+        ecdsa_chip: ECDSAVerifyChip,
+        is_valid: bool = True,
+    ) -> None:
+        self.ecdsa_chip = ecdsa_chip
+        self.pub_key_x_bytes = ecdsa_chip.pub_key_x_bytes
+        self.pub_key_y_bytes = ecdsa_chip.pub_key_y_bytes
+        self.msg_hash_bytes = ecdsa_chip.msg_hash_bytes
+
+        # table
+        self.msg_hash = msg_hash
+        self.sig_v = FQ(int.from_bytes(self.ecdsa_chip.sig_v.le_bytes, "little"))
+        self.sig_r = Word(int.from_bytes(self.ecdsa_chip.sig_r.le_bytes, "little"))
+        self.sig_s = Word(int.from_bytes(self.ecdsa_chip.sig_s.le_bytes, "little"))
+        self.recovered_addr = address
+        self.is_valid = is_valid
+
+        self.pub_key_hash = pub_key_hash
+
+    @classmethod
+    def assign(
+        cls,
+        signature: KeyAPI.Signature,
+        pub_key: KeyAPI.PublicKey,
+        msg_hash: bytes,
+        is_valid: bool = True,
+    ):
+        pub_key_hash = keccak(pub_key.to_bytes())
+        self_pub_key_hash = pub_key_hash
+        self_address = FQ(int.from_bytes(pub_key_hash[-20:], "big"))
+        self_msg_hash = Word(int.from_bytes(msg_hash, "big"))
+        self_ecdsa_chip = ECDSAVerifyChip.assign(signature, pub_key, msg_hash)
+        return cls(self_pub_key_hash, self_address, self_msg_hash, self_ecdsa_chip, is_valid)
+
+    def verify(self, keccak_table: KeccakTable, keccak_randomness: FQ, assert_msg: str):
+        # 0. Copy constraints between pub_key, msg_hash and signature of this chip
+        # and the ones in ECDSA chip
+        assert self.pub_key_x_bytes == self.ecdsa_chip.pub_key_x_bytes
+        assert self.pub_key_y_bytes == self.ecdsa_chip.pub_key_y_bytes
+        assert self.msg_hash_bytes == self.ecdsa_chip.msg_hash_bytes
+        assert self.sig_r.int_value() == int.from_bytes(self.ecdsa_chip.sig_r.le_bytes, "little")
+        assert self.sig_s.int_value() == int.from_bytes(self.ecdsa_chip.sig_s.le_bytes, "little")
+
+        # 1. Constrain v to be equal 0 or 1
+        assert self.sig_v == 0 or self.sig_v == 1
+
+        # 2. Verify that keccak(pub_key_bytes) = pub_key_hash by keccak table
+        # lookup, where pub_key_bytes is built from the pub_key in the
+        # ecdsa_chip
+        pub_key_bytes = self.pub_key_x_bytes + self.pub_key_y_bytes
+        keccak_table.lookup(
+            True,
+            RLC(pub_key_bytes, keccak_randomness, n_bytes=64).expr(),
+            FQ(64),
+            Word(self.pub_key_hash),
+            assert_msg,
+        )
```

Comment:
You're right! I just found we need to reverse `pub_key_bytes` everytime we use it so I store it in big-endian. However, I didn't realize public key in in little-endian.

Wondering if public key is in little-endian, why we convert into big-endian [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/tx_circuit.py#L217-L223) in tx_circuit? I guess it's also related to https://github.com/privacy-scaling-explorations/zkevm-specs/pull/495#discussion_r1372517254. Why do we reverse it in `keecak_table`?

Reply:
fixed in [80e97bd](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/495/commits/80e97bd46ad1de33af8cb26b2649b4d85e8cc9bf)

Reply:
> why we convert into big-endian [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/src/zkevm_specs/tx_circuit.py#L217-L223) in tx_circuit?
Because that's how pk hash is generated? It hashes the public key in big-endian

> hy do we reverse it in `keecak_table`?
Because the hash is seen as big-endian word, to RLC it we need to turn it into little-endian